controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -0,0 +1,480 @@
+package watcher
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/linkerd/linkerd2/controller/k8s"
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/client-go/tools/cache"
+)
+
+const (
+	kubeSystem = "kube-system"
+	podIPIndex = "ip"
+)
+
+// TODO: prom metrics for all the queues/caches
+// https://github.com/linkerd/linkerd2/issues/2204
+
+type (
+	Address struct {
+		Ip        string
+		Port      Port
+		Pod       *corev1.Pod
+		OwnerName string
+		OwnerKind string
+	}
+
+	PodSet = map[PodID]Address
+
+	// endpointsWatcher watches all endpoints and services in the Kubernetes
+	// cluster.  Listeners can subscribe to a particular service and port and
+	// endpointsWatcher will publish the address set and all future changes for
+	// that service:port.
+	EndpointsWatcher struct {
+		publishers   map[ServiceID]*servicePublisher
+		publishersMu sync.RWMutex // This mutex protects modifcation of the map itself.
+		k8sAPI       *k8s.API
+
+		log *logging.Entry
+	}
+
+	// servicePublisher represents a service along with a port number.  Multiple
+	// listeners may be subscribed to a servicePublisher.  servicePublisher maintains the
+	// current state of the address set and publishes diffs to all listeners when
+	// updates come from either the endpoints API or the service API.
+	servicePublisher struct {
+		id     ServiceID
+		log    *logging.Entry
+		k8sAPI *k8s.API
+
+		ports map[Port]*portPublisher
+		// All access to the servicePublisher and its portPublishers is explicitly synchronized by
```

Comment:
I went with one mutex for an entire servicePublisher which guards all operations on its portPublishers.  Most updates to a servicePublisher will affect all of its portPublishers (e.g. an endpoints update) so it seemed simpler to lock them all at once.  This is contrast to how each servicePublisher has a distinct mutex since updates coming into the endpointsWatcher will usually only affect a single servicePublisher.